### PR TITLE
Add structured way to adjust raw system-probe config

### PR DESCRIPTION
**Pull request by brycekahle**
### What does this PR do?

Adjusts the raw system-probe config as early as possible, using structured helpers.

### Motivation

Allows other features like inventories to get the accurate state of product/features.

### Additional Notes

The intent is to try and manipulate the raw config where possible, and do as much of a 1-to-1 translation from raw config to configuration object fields. This should reduce confusion.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/state.go [hunk 1]
```diff
@@ -435,7 +435,7 @@ func (ns *networkState) storeClosedConnections(conns []ConnectionStats) {
 				continue
 			}
 
-			if len(client.closedConnections) >= ns.maxClosedConns {
+			if uint(len(client.closedConnections)) >= ns.maxClosedConns {
```
Comment (hmahmood) [inline pkg/network/state.go:438]: Can we support maxClosedConns > maxInt?

File: cmd/system-probe/config/ns.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package config
+
+import "strings"
+
+func spNS(k ...string) string {
```
Comment (leeavital) [inline cmd/system-probe/config/ns.go:10]: I found these abbreviations a bit inscrutable without foreknowledge of what the products were.  If it's not too verbose, it might be helpful to spell them out.

File: cmd/system-probe/config/adjust.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	if val > max {
+		log.Warnf("limiting configuration key `%s` to maximum value `%d`", key, max)
+		cfg.Set(key, max)
+	}
+}
+
+func limitMaxInt64(cfg config.Config, key string, max int64) {
+	val := cfg.GetInt64(key)
+	if val > max {
+		log.Warnf("limiting configuration key `%s` to maximum value `%d`", key, max)
```
Comment (leeavital) [inline cmd/system-probe/config/adjust.go:95]:
```suggestion
		log.Warnf("%s was configured to %d which exceeds its maximum. %d will be used", key, val, max)
```

File: cmd/system-probe/config/adjust.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func deprecationMessage(oldkey, newkey string) string {
+	return fmt.Sprintf("configuration key `%s` is deprecated, use `%s` instead", oldkey, newkey)
+}
+
+func limitMaxInt(cfg config.Config, key string, max int) {
+	val := cfg.GetInt(key)
+	if val > max {
+		log.Warnf("limiting configuration key `%s` to maximum value `%d`", key, max)
```
Comment (leeavital) [inline cmd/system-probe/config/adjust.go:87]:
```suggestion
		log.Warnf("%s was configured to %d which exceeds its maximum. %d will be used", key, val, max)
```

**Review by leeavital — APPROVED** [review]
Reviewed for NPM, thanks for doing this

**Review by leeavital — APPROVED** [review]
Reviewed for NPM, thanks for doing this

File: cmd/system-probe/config/ns.go (hunk 2 above)
Comment (brycekahle) [reply cmd/system-probe/config/ns.go:10]: The point is to keep them less verbose. I will add comments.

File: pkg/network/state.go (hunk 1 above)
Comment (brycekahle) [reply pkg/network/state.go:438]: The type was changed to `uint` because of the default being applied which was already a `uint`.

Comment (hmahmood) [reply pkg/network/state.go:438]: Can we change the default to `int`? There is no way we can support > maxInt entries right?

Comment (brycekahle) [reply pkg/network/state.go:438]: viper doesn't support `uint`, so we are actually casting from `int` to `uint` right away with `uint(cfg.GetInt(join(spNS, "max_closed_connections_buffered")))`. So I don't think it will ever actually be over max int. I can change them both to `int` though.

File: cmd/system-probe/config/adjust_npm.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	if runtime.GOOS == "windows" {
+		validateInt(cfg, spNS("closed_connection_flush_threshold"), 0, func(v int) error {
+			if v < 1024 {
+				return fmt.Errorf("closed connection notification threshold set to invalid value %d. resetting to default", v)
+			}
+			return nil
+		})
+	}
+
+	validateInt64(cfg, spNS("max_tracked_connections"), defaultMaxTrackedConnections, func(v int64) error {
```
Comment (hmahmood) [inline cmd/system-probe/config/adjust_npm.go:48]: Shouldn't this be something like `MaxInt` or `MaxUint32`?

Comment (brycekahle) [reply cmd/system-probe/config/adjust_npm.go:48]: `Int64` is the type, not the valid range.

Comment (brycekahle) [reply cmd/system-probe/config/adjust_npm.go:48]: I can add a check for the maximum too, if you like.

Comment (hmahmood) [reply cmd/system-probe/config/adjust_npm.go:48]: Yes that is what I meant.

Comment (brycekahle) [reply cmd/system-probe/config/adjust_npm.go:49]: Added

File: cmd/system-probe/config/adjust.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+func deprecateString(cfg config.Config, oldkey string, newkey string) {
+	deprecateCustom(cfg, oldkey, newkey, func(cfg config.Config) interface{} {
+		return cfg.GetString(oldkey)
+	})
+}
+
+func deprecateCustom(cfg config.Config, oldkey string, newkey string, getFn func(config.Config) interface{}) {
+	if cfg.IsSet(oldkey) && !cfg.IsSet(newkey) {
+		log.Info(deprecationMessage(oldkey, newkey))
```
Comment (guyarb) [inline cmd/system-probe/config/adjust.go:88]: IMO, that should be a warn level

File: cmd/system-probe/config/adjust.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+func deprecateString(cfg config.Config, oldkey string, newkey string) {
+	deprecateCustom(cfg, oldkey, newkey, func(cfg config.Config) interface{} {
+		return cfg.GetString(oldkey)
+	})
+}
+
+func deprecateCustom(cfg config.Config, oldkey string, newkey string, getFn func(config.Config) interface{}) {
```
Comment (guyarb) [inline cmd/system-probe/config/adjust.go:86]: let's document the function

File: cmd/system-probe/config/adjust.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	if val > max {
+		log.Warnf("configuration key `%s` was set to `%d`, using maximum value `%d` instead", key, val, max)
+		cfg.Set(key, max)
+	}
+}
+
+func limitMaxInt64(cfg config.Config, key string, max int64) {
+	val := cfg.GetInt64(key)
+	if val > max {
+		log.Warnf("configuration key `%s` was set to `%d`, using maximum value `%d` instead", key, val, max)
```
Comment (guyarb) [inline cmd/system-probe/config/adjust.go:108]: nit: can use `%q` instead of `'%s'`
```suggestion
		log.Warnf("configuration key %q was set to `%d`, using maximum value `%d` instead", key, val, max)
```

File: cmd/system-probe/config/adjust.go (hunk 7 above)
Comment (guyarb) [reply cmd/system-probe/config/adjust.go:86]: generally, let's document all functions

**Review by hmahmood — APPROVED** [review]

File: cmd/system-probe/config/adjust.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+			cfg.Set(key, defaultVal)
+		}
+	} else {
+		cfg.Set(key, defaultVal)
+	}
+}
+
+func validateInt64(cfg config.Config, key string, defaultVal int64, fn func(int64) error) {
+	if cfg.IsSet(key) {
+		if err := fn(cfg.GetInt64(key)); err != nil {
```
Comment (guyarb) [inline cmd/system-probe/config/adjust.go:65]: what's the intention of `fn`?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: cmd/system-probe/config/adjust.go (hunk 8 above)
Comment (brycekahle) [reply cmd/system-probe/config/adjust.go:108]: I'm using backticks, not quotes

File: cmd/system-probe/config/adjust.go (hunk 9 above)
Comment (brycekahle) [reply cmd/system-probe/config/adjust.go:65]: A custom validation function

Comment (guyarb) [reply cmd/system-probe/config/adjust.go:65]: let's document or rename, it is not clear from the function

File: cmd/system-probe/config/adjust.go (hunk 8 above)
Comment (guyarb) [reply cmd/system-probe/config/adjust.go:108]: what's the impact? just formatting?

Comment (brycekahle) [reply cmd/system-probe/config/adjust.go:108]: Less likely to have issues with nested double quotes. Also results in a nicer copied error message in most contexts.

File: cmd/system-probe/config/adjust.go (hunk 6 above)
Comment (brycekahle) [reply cmd/system-probe/config/adjust.go]: Changed

File: cmd/system-probe/config/adjust.go (hunk 7 above)
Comment (brycekahle) [reply cmd/system-probe/config/adjust.go:97]: Done

File: pkg/network/usm/kafka_monitor_test.go [hunk 10]
```diff
@@ -359,6 +360,7 @@ func TestKafkaProtocolParsing(t *testing.T) {
 			},
 			teardown: kafkaTeardown,
 			configuration: func() *config.Config {
+				_, _ = sysconfig.New("/doesnotexist")
 				cfg := config.New()
```
Comment (guyarb) [inline pkg/network/usm/kafka_monitor_test.go:364]: is that mandatory to have? if so, let's use `testConfig` everywhere

**Review by guyarb — CHANGES_REQUESTED** [review]

File: cmd/system-probe/config/adjust.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+// deprecateString sets `newkey` to the value at `oldkey` and logs a deprecation message,
+// but only if `oldkey` is set and `newkey` is not set.
+func deprecateString(cfg config.Config, oldkey string, newkey string) {
+	deprecateCustom(cfg, oldkey, newkey, func(cfg config.Config) interface{} {
+		return cfg.GetString(oldkey)
+	})
+}
+
+// deprecateString sets `newkey` to the value obtained from `getFn` and logs a deprecation message,
```
Comment (guyarb) [inline cmd/system-probe/config/adjust.go:95]:
```suggestion
// deprecateCustom sets `newkey` to the value obtained from `getFn` and logs a deprecation message,
```

File: pkg/network/usm/kafka_monitor_test.go (hunk 10 above)
Comment (brycekahle) [reply pkg/network/usm/kafka_monitor_test.go:364]: Since the adjustments are happening in `cmd/system-probe/config` now, that logic needs to happen first. There is a good chance most of these tests don't care, but I did it just to be sure.

TBH, I don't love how easy it is to do the wrong thing, so I'm thinking about ways to improve this before I merge.

Comment (guyarb) [reply pkg/network/usm/kafka_monitor_test.go:364]: My concern is that people won't be aware of that change (if that's required), and easily they can forget about that, and spend a lot of time just because they forgot to add this line

Comment (brycekahle) [reply pkg/network/usm/kafka_monitor_test.go:364]: Yeah, that is what I meant by "I don't love how easy it is to do the wrong thing".

File: pkg/network/config/config.go [hunk 12]
```diff
@@ -270,22 +263,22 @@ func New() *Config {
...
 		UDPConnTimeout:    defaultUDPTimeoutSeconds * time.Second,
 		UDPStreamTimeout:  defaultUDPStreamTimeoutSeconds * time.Second,
 
 		OffsetGuessThreshold:           uint64(cfg.GetInt64(join(spNS, "offset_guess_threshold"))),
 		ExcludedSourceConnections:      cfg.GetStringMapStringSlice(join(spNS, "source_excludes")),
 		ExcludedDestinationConnections: cfg.GetStringMapStringSlice(join(spNS, "dest_excludes")),
 
-		MaxTrackedConnections:          uint(cfg.GetInt(join(spNS, "max_tracked_connections"))),
-		MaxClosedConnectionsBuffered:   cfg.GetInt(join(spNS, "max_closed_connections_buffered")),
+		MaxTrackedConnections:          uint32(cfg.GetInt64(join(spNS, "max_tracked_connections"))),
```
Comment (usamasaqib) [inline pkg/network/config/config.go:280]: Why read this as an `int64` to begin with?

Comment (brycekahle) [reply pkg/network/config/config.go:280]: So it can handle the full range of `uint32`. Our version of viper doesn't currently support uint, but I'm asking to get it added.

File: pkg/network/protocols/events/consumer_test.go [hunk 13]
```diff
@@ -37,6 +38,7 @@ func TestConsumer(t *testing.T) {
 	}
 
 	const numEvents = 100
+	_, _ = sysconfig.New("/doesnotexist")
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/protocols/events/consumer_test.go:41]: > replace direct access to ddconfig.SystemProbe

This would be even more code change I think. I'm exploring a different way to manage this, so you cannot make a mistake.

File: pkg/network/config/config_test.go [hunk 14]
```diff
@@ -443,6 +447,8 @@ func TestMaxHTTPStatsBuffered(t *testing.T) {
 		newConfig(t)
 		t.Setenv("DD_SYSTEM_PROBE_NETWORK_MAX_HTTP_STATS_BUFFERED", "50000")
 
+		_, err := sysconfig.New("")
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/config/config_test.go:450]: Loading the config does not override environment variables. Rather, it uses them.

File: cmd/system-probe/config/adjust.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+// but only if `oldkey` is set and `newkey` is not set.
+func deprecateCustom(cfg config.Config, oldkey string, newkey string, getFn func(config.Config) interface{}) {
+	if cfg.IsSet(oldkey) && !cfg.IsSet(newkey) {
```
Comment (guyarb) [inline cmd/system-probe/config/adjust.go:98]: I do think we should log the deprecation warning if `cfg.IsSet(oldkey) = true` (regardless of `cfg.IsSet(newkey)`)

File: pkg/security/config/config.go [hunk 16]
```diff
@@ -181,11 +181,6 @@ func NewConfig() (*Config, error) {
 }
 
 func NewRuntimeSecurityConfig() (*RuntimeSecurityConfig, error) {
-	dumpDuration := time.Duration(coreconfig.SystemProbe.GetInt("runtime_security_config.activity_dump.cgroup_dump_timeout")) * time.Minute
```
Comment (paulcacheux) [inline pkg/security/config/config.go:184]: Why was this removed ?

File: pkg/security/config/config.go [hunk 17]
```diff
@@ -271,16 +266,6 @@ func (c *RuntimeSecurityConfig) IsRuntimeEnabled() bool {
 
 // sanitize ensures that the configuration is properly setup
 func (c *RuntimeSecurityConfig) sanitize() error {
-	// if runtime is enabled then we force fim
-	if c.RuntimeEnabled {
```
Comment (paulcacheux) [inline pkg/security/config/config.go:275]: Why was this removed ?

**Review by paulcacheux — CHANGES_REQUESTED** [review]

File: pkg/security/config/config.go (hunk 16 above)
Comment (brycekahle) [reply pkg/security/config/config.go:184]: Both were moved to https://github.com/DataDog/datadog-agent/pull/17188/files#diff-9e22888ba45a2afd83b1a45a4a670f3697a07aa9860351f01bc252994cd0258d

**Review by paulcacheux — APPROVED** [review]

**Review by usamasaqib — APPROVED** [review]

File: pkg/network/tracer/gateway_lookup.go [hunk 18]
```diff
@@ -92,7 +92,7 @@ func newGatewayLookup(config *config.Config) *gatewayLookup {
 	}
 
 	routeCacheSize := maxRouteCacheSize
-	if config.MaxTrackedConnections <= uint(maxRouteCacheSize) {
+	if int(config.MaxTrackedConnections) <= maxRouteCacheSize {
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/tracer/gateway_lookup.go]: Fixed

**Review by guyarb — APPROVED** [review]

File: pkg/network/usm/kafka_monitor_test.go (hunk 10 above)
Comment (brycekahle) [reply pkg/network/usm/kafka_monitor_test.go]: I made it harder to "use wrong", but I think long-term we will want the configs to be components. I started exploring that, but it is a much larger code change.

**Review by kkhor-datadog — APPROVED** [review]
Looks good for Processes, the only file we explicitly own is a unit test, and the handling of the system probe process module config looks unchanged
